File: rename_guides.py
import argparse
import os
import re
import sys
from pathlib import Path

try:
    from PyPDF2 import PdfReader
except ImportError:
    PdfReader = None

PATIENT_PATTERNS = [
    r"Nome do Beneficiário\s*[:\-]?\s*(.+?)\s*(?:\r?\n|$)",
    r"Nome do Titular\s*[:\-]?\s*(.+?)\s*(?:\r?\n|$)",
    r"Nome do paciente\s*[:\-]?\s*(.+?)\s*(?:\r?\n|$)",
    r"Paciente\s*[:\-]?\s*(.+?)\s*(?:\r?\n|$)",
    r"Beneficiário\s*[:\-]?\s*(.+?)\s*(?:\r?\n|$)",
]
GUIDE_PATTERNS = [
    r"Guia\s*(?:N(?:º|o|º?)|\#)?\s*[:\-]?\s*(\d+)",
    r"N[oº] da guia\s*[:\-]?\s*(\d+)",
    r"Número da guia\s*[:\-]?\s*(\d+)",
    r"Guia TISS\s*[:\-]?\s*(\d+)",
]

INVALID_FILENAME_CHARS = r'[<>:"/\\|?*]'


def extract_text_from_pdf(pdf_path: Path) -> str:
    if PdfReader is None:
        raise RuntimeError("PyPDF2 is not installed. Install it with 'pip install -r requirements.txt'.")
    reader = PdfReader(str(pdf_path))
    text_parts = []
    for page in reader.pages:
        try:
            page_text = page.extract_text() or ""
        except Exception:
            page_text = ""
        text_parts.append(page_text)
    return "\n".join(text_parts)


def normalize_name(value: str) -> str:
    value = value.strip()
    value = re.sub(r"\s+", "_", value)
    value = re.sub(INVALID_FILENAME_CHARS, "", value)
    return value


def find_patient_names(text: str) -> list[str]:
    results = []
    for pattern in PATIENT_PATTERNS:
        for match in re.finditer(pattern, text, flags=re.IGNORECASE):
            patient = match.group(1).strip()
            if patient and patient not in results:
                results.append(patient)
    return results


def find_guide_numbers(text: str) -> list[str]:
    results = []
    for pattern in GUIDE_PATTERNS:
        for match in re.finditer(pattern, text, flags=re.IGNORECASE):
            guide = match.group(1).strip()
            if guide and guide not in results:
                results.append(guide)
    return results


def build_renamed_filename(patient: str, guide: str, original_path: Path) -> str:
    safe_patient = normalize_name(patient)
    safe_guide = normalize_name(guide)
    if not safe_patient:
        safe_patient = original_path.stem
    if not safe_guide:
        safe_guide = "guia"
    return f"{safe_patient}_{safe_guide}{original_path.suffix}"


def select_pair(patients: list[str], guides: list[str], requested_guide: str | None = None) -> tuple[str, str] | None:
    if requested_guide:
        for guide in guides:
            if guide == requested_guide:
                idx = guides.index(guide)
                if idx < len(patients):
                    return patients[idx], guide
                return patients[0], guide
    if patients and guides:
        pair_count = min(len(patients), len(guides))
        return patients[0], guides[0]
    if patients:
        return patients[0], guides[0] if guides else ("", "")
    if guides:
        return ("", guides[0])
    return None


def rename_pdf(input_path: Path, output_dir: Path | None = None, guide_number: str | None = None) -> Path:
    text = extract_text_from_pdf(input_path)
    patients = find_patient_names(text)
    guides = find_guide_numbers(text)
    pair = select_pair(patients, guides, request_guide=guide_number)

    if pair is None:
        raise ValueError("Não foi possível extrair o nome do paciente ou número da guia do PDF.")

    patient, guide = pair
    if not patient:
        patient = input_path.stem
    if not guide:
        guide = "guia"

    new_filename = build_renamed_filename(patient, guide, input_path)
    output_dir = output_dir or input_path.parent
    output_path = output_dir / new_filename

    if output_path.exists() and output_path.samefile(input_path):
        return output_path

    input_path.replace(output_path)
    return output_path


def main() -> int:
    parser = argparse.ArgumentParser(description="Renomeia PDFs de guias com base no nome da paciente e número da guia.")
    parser.add_argument("pdfs", type=Path, nargs="+", help="Caminho(s) para o arquivo(s) PDF.")
    parser.add_argument("--output-dir", type=Path, help="Diretório de saída para os PDFs renomeados.")
    parser.add_argument("--guide", dest="guide_number", help="Número da guia para selecionar o paciente correto quando existem vários.")
    args = parser.parse_args()

    had_error = False
    for pdf_path in args.pdfs:
        if not pdf_path.exists():
            print(f"Arquivo não encontrado: {pdf_path}")
            had_error = True
            continue

        try:
            output_path = rename_pdf(pdf_path, args.output_dir, args.guide_number)
            print(f"PDF renomeado com sucesso: {output_path}")
        except Exception as exc:
            print(f"Erro ao processar {pdf_path}: {exc}")
            had_error = True

    return 1 if had_error else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rename_guides.js
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import pdf from "pdf-parse";

const PATTERNS = {
  patient: [
    /Nome do Beneficiário\s*[:\-]?\s*\n?\s*(.+?)\s*(?:\r?\n|$)/gi,
    /Nome do Titular\s*[:\-]?\s*\n?\s*(.+?)\s*(?:\r?\n|$)/gi,
    /Nome do paciente\s*[:\-]?\s*\n?\s*(.+?)\s*(?:\r?\n|$)/gi,
    /Paciente\s*[:\-]?\s*\n?\s*(.+?)\s*(?:\r?\n|$)/gi,
    /Beneficiário\s*[:\-]?\s*\n?\s*(.+?)\s*(?:\r?\n|$)/gi,
  ],

  guide: [
    /Número da Guia no Prestador\s*[:\-]?\s*\n?\s*(\d+)/gi,
    /Número Guia Prestador\s*[:\-]?\s*\n?\s*(\d+)/gi,
    /Número da guia principal\s*[:\-]?\s*\n?\s*(\d+)/gi,
    /N[ºo]\s*Guia Operadora\s*[:\-]?\s*\n?\s*(\d+)/gi,
    /Guia TISS\s*[:\-]?\s*\n?\s*(\d+)/gi,
    /Número da guia\s*[:\-]?\s*\n?\s*(\d+)/gi,
    /Guia\s*(?:N(?:º|o|º?)|#)?\s*[:\-]?\s*\n?\s*(\d+)/gi,
  ],
};

const INVALID_FILENAME_CHARS = /[<>:"/\\|?*]/g;

async function extractTextFromPdf(filePath) {
  const data = await fs.promises.readFile(filePath);
  const parsed = await pdf(data);

  return normalizeText(parsed.text || "");
}

function normalizeText(text) {
  return text
    .replace(/\r/g, "")
    .replace(/[ \t]+/g, " ")
    .replace(/\n\s+/g, "\n")
    .replace(/\n{2,}/g, "\n")
    .trim();
}

function findFirstMatch(text, patterns) {
  for (const pattern of patterns) {
    const match = pattern.exec(text);
    pattern.lastIndex = 0;

    if (match?.[1]) {
      return match[1].trim();
    }
  }

  return "";
}

function findByLabel(text, label) {
  const lines = text
    .split("\n")
    .map(line => line.trim())
    .filter(Boolean);

  const index = lines.findIndex(line =>
    line.toLowerCase().includes(label.toLowerCase())
  );

  return index >= 0 && lines[index + 1]
    ? lines[index + 1].trim()
    : "";
}

function extractData(text) {
  const patient =
    findFirstMatch(text, PATTERNS.patient) ||
    findByLabel(text, "Nome do Beneficiário") ||
    findByLabel(text, "Paciente");

  const guide =
    findFirstMatch(text, PATTERNS.guide) ||
    findByLabel(text, "Número da Guia no Prestador") ||
    findByLabel(text, "Número da guia");

  return { patient, guide };
}

function sanitizeFilename(value) {
  return value
    .trim()
    .replace(/\s+/g, "_")
    .replace(INVALID_FILENAME_CHARS, "")
    .slice(0, 180);
}

function buildRenamedFilename(originalPath, patient, guide) {
  const baseName = path.basename(originalPath, path.extname(originalPath));

  const safePatient = sanitizeFilename(patient || baseName);
  const safeGuide = sanitizeFilename(guide || "guia");

  return `${safePatient}_${safeGuide}${path.extname(originalPath)}`;
}

async function renamePdf(inputPath, outputDir = null) {
  const text = await extractTextFromPdf(inputPath);
  const { patient, guide } = extractData(text);

  if (!patient && !guide) {
    throw new Error("Não foi possível extrair nome do paciente ou número da guia.");
  }

  const newFilename = buildRenamedFilename(inputPath, patient, guide);
  const targetDir = outputDir || path.dirname(inputPath);
  const outputPath = path.join(targetDir, newFilename);

  await fs.promises.mkdir(targetDir, { recursive: true });

  if (path.resolve(inputPath) !== path.resolve(outputPath)) {
    await fs.promises.rename(inputPath, outputPath);
  }

  return outputPath;
}

function parseArgs(argv) {
  const args = {
    pdfs: [],
    outputDir: null,
  };

  for (let i = 2; i < argv.length; i++) {
    const arg = argv[i];

    if (arg === "--output-dir" && argv[i + 1]) {
      args.outputDir = path.resolve(argv[++i]);
    } else if (arg.startsWith("--")) {
      throw new Error(`Argumento desconhecido: ${arg}`);
    } else {
      args.pdfs.push(arg);
    }
  }

  if (!args.pdfs.length) {
    throw new Error(
      "Uso: node rename_guides.js arquivo1.pdf [arquivo2.pdf ...] [--output-dir pasta]"
    );
  }

  return args;
}

async function main() {
  try {
    const { pdfs, outputDir } = parseArgs(process.argv);

    let hasErrors = false;

    for (const file of pdfs) {
      try {
        const resolvedPath = path.resolve(file);

        if (!fs.existsSync(resolvedPath)) {
          console.error(`Arquivo não encontrado: ${resolvedPath}`);
          hasErrors = true;
          continue;
        }

        const outputPath = await renamePdf(resolvedPath, outputDir);
        console.log(`PDF renomeado: ${outputPath}`);
      } catch (error) {
        console.error(`Erro ao processar ${file}: ${error.message}`);
        hasErrors = true;
      }
    }

    process.exit(hasErrors ? 1 : 0);
  } catch (error) {
    console.error(error.message);
    process.exit(1);
  }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  main();
}